### PR TITLE
Feature/answers

### PR DESCRIPTION
**Pull request by siddharth-vinsol**
* User can answer other user’s questions(markdown) from the question’s show page.
* User should be notified by email when an answer is posted to their question.
* Answers are displayed in descending upvotes
* Other users can upvote downvote answers using ajax

### CODE REVIEW

File: app/models/vote.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class Vote < ApplicationRecord
+  belongs_to :voteable, polymorphic: true
+
+  enum vote: {
```
Comment (shubhamjuneja1) [inline app/models/vote.rb:2]: use consistant approach

use enum first
then validation
then associations

File: app/controllers/questions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
+class QuestionsController < ApplicationController
+  before_action :set_question, only: [:publish, :edit, :update, :destroy]
+  before_action :set_question_with_answers, only: [:show]
+  before_action :validate_current_user_resource, only: [:publish, :edit, :update, :destroy]
+  skip_before_action :authorize, only: [:show]
+
+  def index
+    @q = Question.where(user: current_user).order('created_at DESC').ransack(params[:q])
```
Comment (shubhamjuneja1) [inline app/controllers/questions_controller.rb]: order(created_at: :desc)

File: app/controllers/questions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    else
+      render :index, status: :unprocessable_entity
+    end
+  end
+
+  def edit
+  end
+
+  def show
+    @new_answer = Answer.new(question: @question)
```
Comment (shubhamjuneja1) [inline app/controllers/questions_controller.rb]: @question.answers.build

File: app/controllers/questions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      redirect_to questions_path, status: :unprocessable_entity
+    end
+  end
+
+  private def question_params
+    params.require(:question).permit(:title, :content, :attachment)
+  end
+
+  private def set_question
+    @resource = @question = Question.find_by_permalink(params[:permalink])
```
Comment (shubhamjuneja1) [inline app/controllers/questions_controller.rb]: better use alias

File: app/controllers/votes_controller.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+class VotesController < ApplicationController
+  before_action :set_voteable, only: [:upvote, :downvote]
+  
+  def upvote
+    @voteable.upvote(current_user.id)
+    render json: { vote_count: @voteable.total_upvotes - @voteable.total_downvotes, type: 'upvote' }
```
Comment (shubhamjuneja1) [inline app/controllers/votes_controller.rb]: @voteable.total_upvotes - @voteable.total_downvotes should be a method from the module

File: app/models/answer.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
+class Answer < ApplicationRecord
+  include VoteHandler
```
Comment (shubhamjuneja1) [inline app/models/answer.rb:2]: act_as_voteable

File: app/mailers/user_mailer.rb [hunk 7]
```diff
@@ -8,4 +8,12 @@ def reset_password(user)
     @user = user
     mail to: user.email, subject: 'Reset password link'
   end
+
+  def answer_posted(question, question_user, answer, answer_user)
+    @question = question
+    @question_user = question_user
+    @answer = answer
+    @answer_user = answer_user
+    mail to: question_user.email, subject: 'Someone posted an answer on your question'
```
Comment (shubhamjuneja1) [inline app/mailers/user_mailer.rb]: subject should be in i18

File: app/mailers/user_mailer.rb [hunk 8]
```diff
@@ -8,4 +8,12 @@ def reset_password(user)
     @user = user
     mail to: user.email, subject: 'Reset password link'
   end
+
+  def answer_posted(question, question_user, answer, answer_user)
```
Comment (shubhamjuneja1) [inline app/mailers/user_mailer.rb]: avoid sending objects in mailer, pass id and load object there
and do we really need to pass everything?
answer_id should work here

File: app/models/answer.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+  after_create_commit :send_answer_posted_mail
+
+  belongs_to :user
+  belongs_to :question
+  has_rich_text :content
+  has_votes
```
Comment (shubhamjuneja1) [inline app/models/answer.rb]: depenedent: :destroy

File: app/models/question.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  validates :title, :permalink, uniqueness: true, allow_blank: true
+
+  before_validation :assign_permalink, on: :create
+
+  def publish
+    update(published_at: Time.current)
+  end
+
+  def update_attachment(new_attachment)
+    attachment.purge_later
```
Comment (shubhamjuneja1) [inline app/models/question.rb]: what if update failed?
you have already removed previous attachment

**Review by shubhamjuneja1 — CHANGES_REQUESTED** [review]

File: app/controllers/questions_controller.rb (hunk 2 above)
Comment (siddharth-vinsol) [reply app/controllers/questions_controller.rb]: rebasing issue, fixed in #3

File: app/models/answer.rb (hunk 6 above)
Comment (siddharth-vinsol) [reply app/models/answer.rb:2]: named as has_votes

File: app/controllers/questions_controller.rb (hunk 4 above)
Comment (siddharth-vinsol) [reply app/controllers/questions_controller.rb]: alias instance variables

File: app/mailers/user_mailer.rb (hunk 8 above)
Comment (siddharth-vinsol) [reply app/mailers/user_mailer.rb]: Load data in mailers using ID

File: app/controllers/questions_controller.rb (hunk 3 above)
Comment (siddharth-vinsol) [reply app/controllers/questions_controller.rb]: discuss

File: app/models/concerns/vote_handler.rb [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    end
+  end
+
+  def handle_total_votes(vote_value, context)
+    if vote_value == 'upvote'
+      case context
+      when :create
+        increment!(:total_upvotes)
+      when :update
+        increment!(:total_upvotes)
```
Comment (shubhamjuneja1) [inline app/models/concerns/vote_handler.rb:39]: use counter_cache

File: app/controllers/questions_controller.rb [hunk 12]
```diff
@@ -61,7 +63,15 @@ def destroy
     params.require(:question).permit(:title, :content, :attachment)
   end
 
+  private def resource
+    @question
```
Comment (shubhamjuneja1) [inline app/controllers/questions_controller.rb:67]: alias

File: app/javascript/scripts/vote_handler.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
+class VoteHandler {
```
Comment (shubhamjuneja1) [inline app/javascript/scripts/vote_handler.js:1]: not reviewing js

File: app/models/concerns/vote_handler.rb (hunk 11 above)
Comment (siddharth-vinsol) [reply app/models/concerns/vote_handler.rb:39]: discuss

**Review by shubhamjuneja1 — APPROVED** [review]

**Pull request merged**